Custom calendar for academic events

huangshawn1 opened the pull request:
AcademicCalendarVC
- got rid of existing calendar table and upcoming collection
- added CalendarTablePairView to show events for current month

CalendarTablePairView
- currently the table and calendar are constrained to be equal height. adjust this ratio?

CalendarView
- custom calendar that can highlight cells based on EventCalendarEntries
  - currently highlighting cells in black if more than one event color - change to a different color?

MonthSelectorView
- horizontally scrolling CollectionView allowing user to select a month
- contains all months from the first month in the set of entries to the last month in the set of entries
  - only show months from current day forward instead?

Other
- fixed a bug in Date+Extension where >= was checking for <=
- added charRepresentation to DayOfWeek

Issues
- small bug sometimes where the month selector won't scroll to the current month immediately
- currently selected year isn't clearly displayed (can still tell by looking at cells in the table), maybe add the year in the design?

Screenshots
![Simulator Screen Shot - iPhone SE (1st generation) - 2021-04-03 at 16 11 07](https://user-images.githubusercontent.com/20230668/113494067-e6eb0080-9499-11eb-8af7-cbfbd6658254.png)
![Simulator Screen Shot - iPhone 12 Pro Max - 2021-04-03 at 16 33 21](https://user-images.githubusercontent.com/20230668/113494137-68db2980-949a-11eb-851b-af04e1394a04.png)
![IMG_2006](https://user-images.githubusercontent.com/20230668/113494138-6d9fdd80-949a-11eb-8378-e7166bb89325.PNG)

objorkman commented:
A couple of usability/design suggestions:
* Mark today's date with some color/underline/border color
* The light-grey color used for days not in the currently viewed month is not light enough, hard to contrast with the regular dates
* Make the highlighted date boxes a bit smaller so when two are adjacent they don't "touch" (like the Figma)
* Not sure how big of a change this is, but I think it's better to still have all the events in the list, and when you scroll the calendar automatically switches month. This would also allow clicking the calendar to scroll the table view.

Currently the table and calendar are constrained to be equal height. adjust this ratio?
    I think calendar shouldn't be scrollable, and shrink table to fit.

objorkman commented:
Also a height issue:
![Simulator Screen Shot - iPhone 12 Pro - 2021-04-10 at 15 12 47](https://user-images.githubusercontent.com/7193442/114285943-de517780-9a0f-11eb-95aa-ec098033406f.png)

KevWho commented:
>> currently the table and calendar are constrained to be equal height. adjust this ratio?

> I think calendar shouldn't be scrollable, and shrink table to fit.

Adding on to Oscar's suggestion, I also think the calendar shouldn't be scrollable, but there might not be a lot of space left over for the table, depending on the screen size. But it's definitely possible to make the calendar take up less space (reduce font size, cell height of the date #'s, etc.). Another option is maybe the table could be taller, and the entire view could scroll, so the user could scroll down to move the calendar out of the way. What do y'all think?

KevWho commented on berkeley-mobile/Calendar/Academic/Calendar/MonthSelectorView.swift:
```diff
@@ -0,0 +1,162 @@
+        monthSelector.leftAnchor.constraint(equalTo: self.leftAnchor).isActive = true
+        monthSelector.rightAnchor.constraint(equalTo: self.rightAnchor).isActive = true
+        monthSelector.bottomAnchor.constraint(equalTo: self.bottomAnchor).isActive = true
+        self.setHeightConstraint(35)
```
Can use `MonthSelectCell.kCellSize.height` instead of 35.

huangshawn1 commented on berkeley-mobile/Calendar/Academic/AcademicCalendarViewController.swift:
```diff
@@ -9,120 +9,40 @@
 
-fileprivate let kCardPadding: UIEdgeInsets = UIEdgeInsets(top: 16, left: 16, bottom: 16, right: 16)
-fileprivate let kViewMargin: CGFloat = 16
+fileprivate let kCardPadding: UIEdgeInsets = UIEdgeInsets(top: 6, left: 6, bottom: 6, right: 6)
+fileprivate let kViewMargin: CGFloat = 6
```
was able to fix that issue by adding an arranged subview instead of adding a subview

huangshawn1 commented on berkeley-mobile/Calendar/Academic/AcademicCalendarViewController.swift:
```diff
@@ -134,70 +54,24 @@ extension AcademicCalendarViewController {
         let card = CardView()
         card.layoutMargins = kCardPadding
-        scrollingStackView.stackView.addArrangedSubview(card)
+        scrollingStackView.addSubview(card)
```
changed back to addArrangedSubview. that's probably better for consistency/in case we need to add other elements to the stack in the future. this fixed the alignment/margin issue

huangshawn1 commented:
- bolded today's date
- made light-gray lighter
- made highlights smaller (80% width and height)
- calendar is no longer scrollable (height resizes to exactly fit all contents)
  - on smaller phones the table was too small. removing spacing/resizing stuff made the contents too small, so I decided to make the CalendarTablePairView scrollable using a ScrollingStackView
- fixed the height issue
- fixed strong reference cycle, used safe to access monthSymbols, used MonthSelectCell.kCellSize.height

- having all events in a table and automatically switching months would be a sizable change, I think it's better to release as-is and modify later if needed

huangshawn1 commented:
![IMG_2025](https://user-images.githubusercontent.com/20230668/114287923-0e087b80-9a20-11eb-9465-a46702f95c67.PNG)
![IMG_2024](https://user-images.githubusercontent.com/20230668/114287926-0fd23f00-9a20-11eb-95b9-3f00df7be30e.PNG)

KevWho reviewed (approved):
Looks like all my comments have been resolved! @objorkman, what do you think about the changes?

Also, would it be possible to bold the current month as well?

objorkman commented:
All the changes look good!
One last question re the following screenshot - Thoughts on restricting the height of the tableview with the events such that the calendar is never fully off the screen (so it would look like the second screenshot). When the view looks like the first screenshot, its a bit hard to get back to the calendar, since the tableview and the entire pairview scrolls.

![Simulator Screen Shot - iPhone 12 Pro - 2021-04-11 at 21 38 58](https://user-images.githubusercontent.com/7193442/114341378-7afa3f00-9b0e-11eb-8e9c-c516df2e8e96.png)
![Simulator Screen Shot - iPhone 12 Pro - 2021-04-11 at 21 40 16](https://user-images.githubusercontent.com/7193442/114341397-86e60100-9b0e-11eb-9c90-42ee89020d99.png)

huangshawn1 commented:
- bolded current month
- made max table view height 70% of view height so the calendar won't ever be entirely off screen
